File: src/PrinterCalibrate.Core/Detection/RingDetector.cs
using OpenCvSharp;

namespace PrinterCalibrate.Core.Detection;

/// <summary>
/// Default ring detector. Thresholds the part against the background (on the HSV value
/// channel, so multi-colour toolpath renders and real scans both segment cleanly), finds the
/// enclosed holes, and keeps the ring centres. Ring holes are separated from the much larger
/// square lattice cells by <see cref="FilterByRadius"/> (a size cluster), so circularity is only
/// a loose gate to drop slivers — real printed/scanned holes are rough (circularity ~0.2–0.8),
/// so a strict threshold would reject them.
/// </summary>
public sealed class RingDetector : IRingDetector
{
    private readonly double _minHoleAreaPx;
    private readonly double _minCircularity;

    public RingDetector(double minHoleAreaPx = 40.0, double minCircularity = 0.20)
    {
        _minHoleAreaPx = minHoleAreaPx;
        _minCircularity = minCircularity;
    }

    public IReadOnlyList<DetectedRing> Detect(Mat image)
    {
        if (image is null || image.Empty())
            throw new ArgumentException("Image is null or empty.", nameof(image));

        using var value = ExtractValueChannel(image);
        using var binary = new Mat();
        Cv2.Threshold(value, binary, 0, 255, ThresholdTypes.Binary | ThresholdTypes.Otsu);

        // Make the part white and the background black, whichever way the contrast falls.
        if (BorderMean(binary) > 127.0)
            Cv2.BitwiseNot(binary, binary);

        using var kernel = Cv2.GetStructuringElement(MorphShapes.Ellipse, new Size(3, 3));
        Cv2.MorphologyEx(binary, binary, MorphTypes.Close, kernel);

        Cv2.FindContours(binary, out Point[][] contours, out HierarchyIndex[] hierarchy,
            RetrievalModes.CComp, ContourApproximationModes.ApproxNone);

        var candidates = new List<DetectedRing>();
        for (int i = 0; i < contours.Length; i++)
        {
            if (hierarchy[i].Parent < 0)
                continue; // only interior contours (holes) can be ring centres

            Point[] contour = contours[i];
            double area = Cv2.ContourArea(contour);
            if (area < _minHoleAreaPx)
                continue;

            double perimeter = Cv2.ArcLength(contour, true);
            if (perimeter <= 0)
                continue;

            double circularity = 4.0 * Math.PI * area / (perimeter * perimeter);
            if (circularity < _minCircularity)
                continue;

            Moments moments = Cv2.Moments(contour);
            if (moments.M00 == 0)
                continue;

            double cx = moments.M10 / moments.M00;
            double cy = moments.M01 / moments.M00;
            double radius = Math.Sqrt(area / Math.PI);
            candidates.Add(new DetectedRing(cx, cy, radius, circularity));
        }

        return FilterByRadius(candidates);
    }

    /// <summary>Drop anything whose radius is far from the population median (stray holes).</summary>
    private List<DetectedRing> FilterByRadius(List<DetectedRing> candidates)
    {
        if (candidates.Count == 0)
            return candidates;

        double median = Median(candidates.Select(c => c.RadiusPx));
        return candidates
            .Where(c => c.RadiusPx >= median * 0.5 && c.RadiusPx <= median * 1.8)
            .ToList();
    }

    private Mat ExtractValueChannel(Mat image)
    {
        if (image.Channels() == 1)
            return image.Clone();

        using var hsv = new Mat();
        Cv2.CvtColor(image, hsv, ColorConversionCodes.BGR2HSV);
        Mat[] channels = Cv2.Split(hsv);
        try
        {
            return channels[2].Clone(); // V = max(B,G,R)
        }
        finally
        {
            foreach (Mat channel in channels)
                channel.Dispose();
        }
    }

    private double BorderMean(Mat binary)
    {
        using Mat top = binary.Row(0);
        using Mat bottom = binary.Row(binary.Rows - 1);
        using Mat left = binary.Col(0);
        using Mat right = binary.Col(binary.Cols - 1);
        return (Cv2.Mean(top).Val0 + Cv2.Mean(bottom).Val0 +
                Cv2.Mean(left).Val0 + Cv2.Mean(right).Val0) / 4.0;
    }

    private double Median(IEnumerable<double> values)
    {
        var sorted = values.OrderBy(v => v).ToList();
        int n = sorted.Count;
        if (n == 0)
            return 0;
        return n % 2 == 1 ? sorted[n / 2] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0;
    }
}

File: web/src/engine/ringDetector.ts
import type { Mat, OpenCv } from './opencv'
import type { DetectedRing } from './types'
import { median } from './math'
import { valueChannel } from './cvUtils'

// Thresholds the image on the HSV value channel (Otsu), finds the enclosed holes, and keeps the ring
// centres (the binary area centroid, immune to over/under extrusion). Ring holes are separated from
// the much larger square lattice cells by a size cluster (radius-median filter), so circularity is
// only a loose gate to drop slivers (real holes are rough, circularity ~0.2 to 0.8).
//
// Which side of the threshold is the part is NOT guessed here. A border statistic proved unreliable:
// a backing sheet that stops short of the scan bed leaves bright scanner-lid margins on the image
// border, flipping the guess and turning dust specks into the only "holes". Instead detection runs
// under BOTH polarities and the caller validates each candidate set against the coupon's known grid
// and orientation marker (model selection against the coupon model), keeping the one that fits.

/** Whether the part is assumed brighter or darker than what is behind it. */
export type RingPolarity = 'bright' | 'dark'

export interface DualDetection {
  bright: DetectedRing[]
  dark: DetectedRing[]
}

// When `masksOut` is passed, the binary mask each polarity ran findContours on (the value channel
// thresholded, oriented part-white, and morphologically closed) is cloned into it so the caller can
// show the user exactly what the detector searched. The caller owns and must delete both masks.
export function detectRingsDual(
  cv: OpenCv,
  image: Mat,
  minHoleAreaPx = 40.0,
  minCircularity = 0.2,
  masksOut?: { bright?: Mat; dark?: Mat },
): DualDetection {
  if (!image || image.empty()) throw new Error('Image is null or empty.')

  // A single-channel image is used directly (no copy); only a colour image allocates a new channel.
  const value = image.channels() === 1 ? image : valueChannel(cv, image)
  const binary = new cv.Mat()
  try {
    cv.threshold(value, binary, 0, 255, cv.THRESH_BINARY + cv.THRESH_OTSU)

    // 'bright' takes the above-threshold pixels as the part; 'dark' the inverse.
    const wantMasks = masksOut !== undefined
    const bright = detectOnBinary(cv, binary, minHoleAreaPx, minCircularity, wantMasks)
    if (masksOut && bright.mask) masksOut.bright = bright.mask
    cv.bitwise_not(binary, binary)
    const dark = detectOnBinary(cv, binary, minHoleAreaPx, minCircularity, wantMasks)
    if (masksOut && dark.mask) masksOut.dark = dark.mask
    return { bright: bright.rings, dark: dark.rings }
  } catch (e) {
    // Don't orphan an already-captured mask when the other polarity's pass throws.
    masksOut?.bright?.delete()
    masksOut?.dark?.delete()
    if (masksOut) {
      delete masksOut.bright
      delete masksOut.dark
    }
    throw e
  } finally {
    if (value !== image) value.delete()
    binary.delete()
  }
}

// Runs the hole search on one part-white binary. Closes small gaps first (on a copy; the input is
// reused for the other polarity), clones the searched mask when asked, then keeps the interior
// contours that pass the area and circularity gates and the radius cluster. The caller owns `mask`.
function detectOnBinary(
  cv: OpenCv,
  partWhite: Mat,
  minHoleAreaPx: number,
  minCircularity: number,
  wantMask: boolean,
): { rings: DetectedRing[]; mask: Mat | null } {
  const closed = new cv.Mat()
  const contours = new cv.MatVector()
  const hierarchy = new cv.Mat()
  let mask: Mat | null = null
  try {
    const kernel = cv.getStructuringElement(cv.MORPH_ELLIPSE, new cv.Size(3, 3))
    cv.morphologyEx(partWhite, closed, cv.MORPH_CLOSE, kernel)
    kernel.delete()

    // Capture the mask before findContours, which can mutate its input.
    if (wantMask) mask = closed.clone()

    // CHAIN_APPROX_SIMPLE drops only collinear boundary points, so contourArea, arcLength, and the
    // moments (all polygon integrals) are unchanged while large lattice-cell contours shrink a lot.
    cv.findContours(closed, contours, hierarchy, cv.RETR_CCOMP, cv.CHAIN_APPROX_SIMPLE)

    const candidates: DetectedRing[] = []
    const count = contours.size()
    for (let i = 0; i < count; i++) {
      // hierarchy rows are [next, prev, firstChild, parent]; read data32S per iteration, since a
      // wasm heap growth mid-loop detaches any TypedArray view captured before it.
      if (hierarchy.data32S[i * 4 + 3] < 0) continue // only interior contours (holes) can be ring centres

      const contour = contours.get(i)
      try {
        const area = cv.contourArea(contour, false)
        if (area < minHoleAreaPx) continue

        const perimeter = cv.arcLength(contour, true)
        if (perimeter <= 0) continue

        const circularity = (4.0 * Math.PI * area) / (perimeter * perimeter)
        if (circularity < minCircularity) continue

        const m = cv.moments(contour, false)
        if (m.m00 === 0) continue

        candidates.push({
          centerX: m.m10 / m.m00,
          centerY: m.m01 / m.m00,
          radiusPx: Math.sqrt(area / Math.PI),
          circularity,
        })
      } finally {
        contour.delete()
      }
    }

    return { rings: filterByRadius(candidates), mask }
  } catch (e) {
    mask?.delete() // don't orphan the captured mask when the contour pass throws
    throw e
  } finally {
    closed.delete()
    contours.delete()
    hierarchy.delete()
  }
}

// Drop anything whose radius is far from the population median (stray holes / lattice cells).
function filterByRadius(candidates: DetectedRing[]): DetectedRing[] {
  if (candidates.length === 0) return candidates
  const med = median(candidates.map((c) => c.radiusPx))
  return candidates.filter((c) => c.radiusPx >= med * 0.5 && c.radiusPx <= med * 1.8)
}

